java-checks/src/main/java/org/sonar/java/checks/RedundantThrowsDeclarationCheck.java

```diff
@@ -43,83 +51,128 @@
   }
 
-  private static boolean isDeclaredMoreThanOnce(AsmClass thrownClass, List<AsmClass> thrownClassCandidates) {
-    int matches = 0;
+  @Override
+  public void leaveNode(Tree tree) {
+    if (!hasSemantic()) {
+      return;
+    }
+    if (tree.is(Tree.Kind.CONSTRUCTOR, Tree.Kind.METHOD)) {
+      MethodTree methodTree = (MethodTree) tree;
+      List<TypeTree> exceptionsTree = new ArrayList<>(methodTree.throwsClauses());
+      checkRuntimeExceptions(methodTree, exceptionsTree);
+      checkRedundantExceptions(methodTree, exceptionsTree);
+      checkOtherExceptions(methodTree, exceptionsTree);
+    }
+    if (tree.is(Tree.Kind.CLASS, Tree.Kind.ENUM, Tree.Kind.INTERFACE, Tree.Kind.CONSTRUCTOR, Tree.Kind.METHOD)) {
+      thrownExceptions.pop();
+    }
+  }
 
-    for (AsmClass thrownClassCandidate : thrownClassCandidates) {
-      if (thrownClass.equals(thrownClassCandidate)) {
-        matches++;
+  private void checkRuntimeExceptions(Tree tree, List<TypeTree> exceptionsTree) {
+    for (int i = exceptionsTree.size() - 1; i >= 0; i--) {
+      TypeTree exceptionTree = exceptionsTree.get(i);
+      Type exceptionType = exceptionTree.symbolType();
+      if (exceptionType.isSubtypeOf("java.lang.RuntimeException")) {
+        addIssue(tree, "Remove the declaration of thrown exception '" + exceptionType.fullyQualifiedName() + "' which is a runtime exception.");
+        exceptionsTree.remove(i);
       }
     }
+  }
 
-    return matches > 1;
+  private void checkRedundantExceptions(MethodTree tree, List<TypeTree> exceptionsTree) {
+    for (int i1 = exceptionsTree.size() - 1; i1 >= 0; i1--) {
+      TypeTree exceptionTree = exceptionsTree.get(i1);
+      Type exceptionType = exceptionTree.symbolType();
+      for (int i2 = i1 - 1; i2 >= 0; i2--) {
+        Type secondExceptionType = exceptionsTree.get(i2).symbolType();
+        if (exceptionType.equals(secondExceptionType) && !exceptionType.symbol().equals(Symbols.unknownSymbol)) {
+          addIssue(tree, "Remove the redundant '" + exceptionType.fullyQualifiedName() + "' thrown exception declaration(s).");
+          exceptionsTree.remove(i1);
+          break;
+        }
+      }
+    }
   }
 
-  private static boolean isSubClassOfAny(AsmClass thrownClass, List<AsmClass> thrownClassCandidates) {
-    for (AsmClass current = thrownClass.getSuperClass(); current != null; current = current.getSuperClass()) {
-      for (AsmClass thrownClassCandidate : thrownClassCandidates) {
-        if (current.equals(thrownClassCandidate)) {
-          return true;
+  private void checkOtherExceptions(MethodTree tree, List<TypeTree> exceptionsTree) {
+    for (int i1 = exceptionsTree.size() - 1; i1 >= 0; i1--) {
+      TypeTree exceptionTree = exceptionsTree.get(i1);
+      Type exceptionType = exceptionTree.symbolType();
+      if (!exceptionType.symbol().equals(Symbols.unknownSymbol)) {
+        if (!checkRelatedExceptions(tree, exceptionTree, exceptionsTree) && shouldCheckExceptionsInBody(tree)
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Merge this if statement with the enclosing one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1066)